refactor OpenMP `parallel_for` implementation to use more native OpenMP constructs

cz4rs opened the pull request:
fixes #4369 

---
Initial benchmarking results: [link](https://github.com/NexGenAnalytics/kokkos/wiki/Benchmarks/)

cz4rs commented on BUILD.md:
```diff
@@ -177,6 +177,9 @@ Options can be enabled by specifying `-DKokkos_ENABLE_X`.
 * Kokkos_ENABLE_PROFILING_LOAD_PRINT
     * Whether to print information about which profiling tools gotloaded
     * BOOL Default: OFF
+* Kokkos_ENABLE_NATIVE_OPENMP
```
This should be either `OFF` by default or (preferably) removed completely.

cz4rs commented on cmake/kokkos_enable_options.cmake:
```diff
@@ -55,6 +55,7 @@ KOKKOS_ENABLE_OPTION(COMPILER_WARNINGS    OFF "Whether to print all compiler war
 KOKKOS_ENABLE_OPTION(PROFILING_LOAD_PRINT OFF "Whether to print information about which profiling tools got loaded")
 KOKKOS_ENABLE_OPTION(TUNING               OFF "Whether to create bindings for tuning tools")
 KOKKOS_ENABLE_OPTION(AGGRESSIVE_VECTORIZATION OFF "Whether to aggressively vectorize loops")
+KOKKOS_ENABLE_OPTION(NATIVE_OPENMP        ON  "Whether to enable refactored OpenMP backend implementation")
```
Enabled temporarily to get the CI feedback.
This should be either `OFF` by default or (preferably) removed completely.

cz4rs commented on benchmarks/benchmark_suite/scripts/run_tests.bash:
```diff
@@ -3,15 +3,9 @@
 # BytesAndFlops
 cd build/bytes_and_flops
 
-USE_CUDA=`grep "_CUDA" KokkosCore_config.h | wc -l`
```
FIXME: revert this

Rombur commented:
It seems that it's a bug in the compiler. Unfortunately, I cannot reproduce it outside of Kokkos. Here is what happens: we have a test  with a `for loop` that looks like `for(int i=0; i<0; ++i)` and for some reason when we use the dynamic scheduler, we enter the loop. Obviously that creates a segfault. The solution is to add 
```C++
     if (m_policy.begin() >= m_policy.end())
           return;
```
before the `#pragma omp parallel for` [here](https://github.com/kokkos/kokkos/pull/4664/files#diff-2918a925908a5355ef92cabbabc7f9e962815febb2a45bcc9ff9d428091fdf65R130). For me all the tests pass wit this change.

dalg24 commented on benchmarks/benchmark_suite/scripts/checkout_repos.bash:
```diff
@@ -3,11 +3,11 @@
+  git checkout develop
+  git pull
+  cd ..
 fi
```
Why did you change?  The intent seemed to be to pull the latest version when the repo had already been cloned

cz4rs replied:
True, that's a change of behavior, but I thought it would be more convenient. You don't need to modify the script when you are running the benchmark on a branch and no surprising pulls happen.

cz4rs replied:
That's a bit different from the other libraries in the script, but I think it should be fine since it's Kokkos that we are benchmarking here.

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -56,6 +56,20 @@
+#define KOKKOS_OPENMP_OPTIONAL_CHUNK_SIZE , m_policy.chunk_size()
+#else
+#define KOKKOS_OPENMP_OPTIONAL_CHUNK_SIZE
+#endif
```
Make sure you `#undef` them

Rombur commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -92,53 +102,92 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
+  typename std::enable_if<std::is_same<typename Policy::schedule_type::type,
+                                       Kokkos::Dynamic>::value>::type
+  execute_parallel() const {
+    if (m_policy.begin() >= m_policy.end()) return;
```
Add a comment that we do this because we hit a compiler bug with nvhpc 11.4.

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -92,53 +102,93 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
     }
   }
 
+  template <class TagType>
+  inline static
+      typename std::enable_if<std::is_same<TagType, void>::value>::type
```
```suggestion
  template <class Enable = WorkTag>
  static std::enable_if_t<std::is_void<WorkTag>::value && std::is_same<Enable, WorkTag>::value>
```
Then at call site `ParallelFor::template exec_work<WorkTag>(m_functor, iWork) -> exec_work(m_functor, iWork);`

dalg24 replied:
Yes, not a bad idea to fix that one too

Rombur commented:
Can we merge this. I have PRs that are almost ready but they conflict with this PR and I am waiting for this code to be merged to finish them.

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
+  }
+
+  template <class Policy>
+  typename std::enable_if<std::is_same<typename Policy::schedule_type::type,
+                                       Kokkos::Dynamic>::value>::type
```
```suggestion
  std::enable_if_t<std::is_same<typename Policy::schedule_type::type, Kokkos::Dynamic>::value>
```
?

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
   }
 
+  template <class Enable = WorkTag>
+  inline static typename std::enable_if_t<std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
```
```suggestion
  inline static std::enable_if_t<std::is_void<WorkTag>::value && std::is_same<Enable, WorkTag>::value>
```
?

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
+  }
+
+  template <class Enable = WorkTag>
+  inline static typename std::enable_if_t<!std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
```
```suggestion
  inline static std::enable_if_t<!std::is_void<WorkTag>::value && std::is_same<Enable, WorkTag>::value>
```
?

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
-#endif
-    for (Member iwork = ibeg; iwork < iend; ++iwork) {
+  template <class Enable = WorkTag>
+  inline static typename std::enable_if_t<!std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
```
```suggestion
  inline static std::enable_if_t<!std::is_void<WorkTag>::value && std::is_same<Enable, WorkTag>::value>
```
?

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
-#endif
-    for (Member iwork = ibeg; iwork < iend; ++iwork) {
+  template <class Enable = WorkTag>
+  inline static typename std::enable_if_t<std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
```
```suggestion
  inline static std::enable_if_t<std::is_void<WorkTag>::value && std::is_same<Enable, WorkTag>::value>
```
?

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
+  inline static typename std::enable_if_t<!std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
+  exec_range(const FunctorType& functor, const Member ibeg, const Member iend) {
+    const Enable t{};
+    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
+    for (auto iwork = ibeg; iwork < iend; ++iwork) {
       functor(t, iwork);
     }
```
```suggestion
    const WorkTag tag{};
    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
    for (auto iwork = ibeg; iwork < iend; ++iwork) {
      functor(tag, iwork);
    }
```

cz4rs commented on benchmarks/benchmark_suite/scripts/run_benchmark.bash:
```diff
@@ -1,14 +1,33 @@
+  local logfile=run-benchmark--$(date +"%Y-%m-%d-%H-%M").log
+
+  cd ${PWD}/kokkos
+  local git_info="KOKKOS_BRANCH: $(git branch --show-current)
```
`git branch --show-current` requires Git 2.22. I can use `rev-parse` here if that's too fresh for some reason.

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
+  }
+
+  template <class Policy>
+  typename std::enable_if<!std::is_same<typename Policy::schedule_type::type,
+                                        Kokkos::Dynamic>::value>::type
```
```suggestion
  std::enable_if_t<!std::is_same<typename Policy::schedule_type::type, Kokkos::Dynamic>::value>
```

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
+#ifndef KOKKOS_INTERNAL_DISABLE_NATIVE_OPENMP
+    execute_parallel<Policy>();
+#else
     enum {
       is_dynamic = std::is_same<typename Policy::schedule_type::type,
                                 Kokkos::Dynamic>::value
     };
```
```suggestion
    constexpr bool is_dynamic = std::is_same<typename Policy::schedule_type::type, Kokkos::Dynamic>::value;
```

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -167,55 +219,84 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
+#ifndef KOKKOS_INTERNAL_DISABLE_NATIVE_OPENMP
+    execute_parallel<Policy>();
+#else
     enum {
       is_dynamic = std::is_same<typename Policy::schedule_type::type,
                                 Kokkos::Dynamic>::value
     };
```
```suggestion
    constexpr bool is_dynamic = std::is_same<typename Policy::schedule_type::type, Kokkos::Dynamic>::value;
```

masterleinad reviewed (commented):
I would prefer if all the changes to the `benchmark` directory would be in a separate pull request.

PhilMiller commented:
retest this please

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -167,55 +216,83 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
+    }
+
+    OpenMPExec::verify_is_master("Kokkos::OpenMP parallel_for");
+#ifndef KOKKOS_INTERNAL_DISABLE_NATIVE_OPENMP
```
Where is this defined?

cz4rs replied:
This is not defined anywhere - it's intended as an advanced option for developers (so you'd have to go with `add_definitions(-DKOKKOS_INTERNAL_DISABLE_NATIVE_OPENMP)` or similar). This was requested not to be exposed as public option, but a temporary fallback.

masterleinad replied:
Ah, OK.

cz4rs commented:
Rebased on top of `develop`, hopefully jenkins likes it this time.

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -71,70 +84,106 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
-#endif
-#endif
-    for (Member iwork = ibeg; iwork < iend; ++iwork) {
+  template <class Enable = WorkTag>
+  inline static std::enable_if_t<std::is_void<WorkTag>::value &&
+                                 std::is_same<Enable, WorkTag>::value>
+  exec_range(const FunctorType& functor, const Member ibeg, const Member iend) {
+    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
+    for (auto iwork = ibeg; iwork < iend; ++iwork) {
       functor(iwork);
     }
   }
```
Why is not just
```suggestion
  inline static exec_range(const FunctorType& functor, const Member ibeg, const Member iend) {
    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
    for (auto iwork = ibeg; iwork < iend; ++iwork) {
      exec_work(functor, iwork);
    }
  }
```

cz4rs replied:
Added that and simplified the implementation for `MDRangePolicy` as well :+1:

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -167,55 +216,83 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
+  }
+
+  template <class Policy>
+  typename std::enable_if<std::is_same<typename Policy::schedule_type::type,
+                                       Kokkos::Dynamic>::value>::type
```
Please use `enable_if_t`

cz4rs commented:
@dalg24 I have addressed all the comments.

Jenkins has two builds failing:
- `OpenMPTarget` (`KokkosCore_UnitTest_OpenMPTarget`)
- `SYCL-OneApi` (`KokkosCore_PerfTestExec`)
I believe these are unrelated to this PR. If so, hopefully this can be merged.
